### PR TITLE
Add FilesystemHelper::create_directory

### PR DESCRIPTION
**Pull request by zmichaels11**
### Changes
* Add `FilesystemHelper::create_directory` (same functionality as C++17 [std::filesystem::create_directory](https://en.cppreference.com/w/cpp/filesystem/create_directory))
* Add unit test validating `Filesystem::create_directory`

### Issues
* N/A (mentioned as required in https://github.com/ros2/rosbag2/pull/185#issuecomment-553975788)

### CODE REVIEW

File: rosbag2_storage/test/rosbag2_storage/test_filesystem_helper.cpp [hunk 1]
```diff
@@ -106,3 +106,23 @@ TEST(FilesystemHelper, get_file_name_returns_empty_string_if_passed_a_path_with_
...
+{
+  const auto bad_path = FilesystemHelper::concat({temporary_dir_path_, "bad_folder"});
+
+  EXPECT_FALSE(FilesystemHelper::file_exists(bad_path));
+
+  const auto path = FilesystemHelper::concat({bad_path, "some_folder"});
+
+  EXPECT_FALSE(FilesystemHelper::create_directory(path));
+  EXPECT_FALSE(FilesystemHelper::file_exists(path));
+}
```
Comment (dabonnie) [inline rosbag2_storage/test/rosbag2_storage/test_filesystem_helper.cpp]: Nit: missing newline?

File: rosbag2_storage/include/rosbag2_storage/filesystem_helper.hpp [hunk 2]
```diff
@@ -141,6 +142,24 @@ class FilesystemHelper
...
+   * The parent directory must already exist.
+   * @param p the path to the new directory to create.
+   * @return true if a directory was created for the directory p.
+   */
+  static bool create_directory(const std::string & p)
+  {
+#ifdef _WIN32
+    _mkdir(p.c_str());
+#else
+    constexpr const int perms_all = S_IRWXU | S_IRWXG | S_IRWXO;  // same as 0777
```
Comment (Karsten1987) [inline rosbag2_storage/include/rosbag2_storage/filesystem_helper.hpp]: so I believe Unix systems create folders with 755 by default. Why 777?

File: rosbag2_storage/test/rosbag2_storage/test_filesystem_helper.cpp [hunk 3]
```diff
@@ -106,3 +106,23 @@ TEST(FilesystemHelper, get_file_name_returns_empty_string_if_passed_a_path_with_
 
   EXPECT_THAT(folder_name, Eq(""));
 }
+
+TEST_F(FilesystemHelperFixture, create_directory_passes_when_parent_exists)
+{
+  const auto path = FilesystemHelper::concat({temporary_dir_path_, "some_folder"});
+
+  EXPECT_TRUE(FilesystemHelper::create_directory(path));
```
Comment (Karsten1987) [inline rosbag2_storage/test/rosbag2_storage/test_filesystem_helper.cpp]: what's happening if that folder existed already? Say you invoke this command multiple times in a row?

File: rosbag2_storage/test/rosbag2_storage/test_filesystem_helper.cpp (hunk 1 above)
Comment (Karsten1987) [inline rosbag2_storage/test/rosbag2_storage/test_filesystem_helper.cpp]: missing return line

Comment (zmichaels11) [reply rosbag2_storage/test/rosbag2_storage/test_filesystem_helper.cpp]: done.

File: rosbag2_storage/include/rosbag2_storage/filesystem_helper.hpp (hunk 2 above)
Comment (zmichaels11) [reply rosbag2_storage/include/rosbag2_storage/filesystem_helper.hpp]: `std::filesystem::create_directory` documentation says it uses 
```c++
static_cast<int>(std::filesystem::perms::all)
```
which maps to 777

**Comment by dirk-thomas** [comment]
That functionality already exists in `rcutils`: https://github.com/ros2/rcutils/blob/4831316a2223de11f731d0497a69911f04ed4976/include/rcutils/filesystem.h#L168

File: rosbag2_storage/test/rosbag2_storage/test_filesystem_helper.cpp (hunk 3 above)
Comment (zmichaels11) [reply rosbag2_storage/test/rosbag2_storage/test_filesystem_helper.cpp]: The test fixture should have deleted the temp folder on tear down.

File: rosbag2_storage/test/rosbag2_storage/test_filesystem_helper.cpp (hunk 1 above)
Comment (zmichaels11) [reply rosbag2_storage/test/rosbag2_storage/test_filesystem_helper.cpp]: Added newline.

**Comment by Karsten1987** [comment]
Thanks for the info. I guess it generally makes sense to merge rosbag2's filessytem helper with `rcutils`.

File: rosbag2_storage/include/rosbag2_storage/filesystem_helper.hpp (hunk 2 above)
Comment (zmichaels11) [reply rosbag2_storage/include/rosbag2_storage/filesystem_helper.hpp]: The design choise to check `file_exists` on return is to match up with C++17's `std::filesystem::create_directory` which returns True if the directory is created.

**Comment by Karsten1987** [comment]
@zmichaels11 I feel we should use the functionality already established in rcutils before introducing the same again here in rosbag2.

**Comment by zmichaels11** [comment]
Yeah, I got mixed up and thought it was FilesystemHelper in rosbag2 that will eventually be replaced by `std::filesystem` with C++17

It looks like this function is already included in `rcpputils` as well

**Pull request closed without merging**